### PR TITLE
Threshold decryption plugin implementation for OCR2.

### PR DESCRIPTION
**Pull request by stchrysa**

### CODE REVIEW

File: go/ocr2/decryptionplugin/decryption.go [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+		ciphertextBytes := request.Ciphertext
+		if err := ciphertext.UnmarshalVerify(ciphertextBytes, dp.publicKey); err != nil {
+			dp.logger.Error("DecryptionReporting Observation: cannot unmarshal and verify the ciphertext, the leader is faulty", commontypes.LogFields{
+				"error":        err,
+				"ciphertextID": request.CiphertextID,
+			})
+			return nil, fmt.Errorf("cannot unmarshal and verify the ciphertext: %w", err)
+		}
+		if dp.specificConfig.Config.LocalRequest {
+			queueCiphertextBytes, err := dp.decryptionQueue.GetCiphertext(request.CiphertextID)
```
Comment (pszal) [inline go/ocr2/decryptionplugin/decryption.go:149]: btw. will the plugin implement how CiphertextIDs are assigned? Or it just requires that IDs are unique per ciphertext?

File: go/ocr2/decryptionplugin/decryption.go [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+				})
+				break
+			}
+		}
+	}
+
+	reportProto := Report{}
+	for id, decrShares := range validDecryptionShares {
+		ciphertext, ok := ciphertexts[id]
+		if !ok {
```
Comment (pszal) [inline go/ocr2/decryptionplugin/decryption.go:272]: seems to me like it is always `ok` here

File: go/ocr2/decryptionplugin/decryption.go [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+			dp.logger.Error("DecryptionReporting Report: there is not ciphertext in the query with matching id, skipping aggregation of decryption shares", commontypes.LogFields{
+				"ciphertextID": id,
+			})
+			continue
+		}
+
+		plaintext, err := ciphertext.Aggregate(decrShares, dp.genericConfig.N)
```
Comment (pszal) [inline go/ocr2/decryptionplugin/decryption.go:279]: maybe put at least a comment that the OCR2 protocol guarantees that `len(decShares)> F`

File: go/ocr2/decryptionplugin/decryption.go [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+// ShouldAcceptFinalizedReport updates the decryption queue.
+// Returns always false as the report will not be transmitted on-chain.
+func (dp *decryptionPlugin) ShouldAcceptFinalizedReport(ctx context.Context, ts types.ReportTimestamp, report types.Report) (bool, error) {
+	dp.logger.Debug("DecryptionReporting ShouldAcceptFinalizedReport: start", commontypes.LogFields{
+		"epoch": ts.Epoch,
+		"round": ts.Round,
+	})
+
+	reportProto := &Report{}
+	err := proto.Unmarshal(report, reportProto)
```
Comment (pszal) [inline go/ocr2/decryptionplugin/decryption.go:348]: inline

File: go/ocr2/decryptionplugin/decryption.go (hunk 1 above)
Comment (stchrysa) [reply go/ocr2/decryptionplugin/decryption.go:149]: The latter. I will add comments in the interface to clarify this.

File: go/ocr2/decryptionplugin/decryption.go [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+		int(dp.specificConfig.Config.RequestCountLimit),
+		int(dp.specificConfig.Config.RequestTotalBytesLimit),
+	)
+
+	queryProto := Query{}
+	for _, request := range decryptionRequests {
+		ciphertext := &tdh2easy.Ciphertext{}
+		err := ciphertext.UnmarshalVerify(request.ciphertext, dp.publicKey)
+		if err != nil {
+			dp.logger.Error("DecryptionReporting Observation: cannot unmarshal the ciphertext, skipping it", commontypes.LogFields{
```
Comment (bolekk) [inline go/ocr2/decryptionplugin/decryption.go:99]: Log message phase should be "Query", not "Observation".

File: go/ocr2/decryptionplugin/config/config_types.proto [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+  uint32 KeyShareIndex = 2;
+}
+
+message ReportingPluginConfig {
+  uint32 maxQueryLengthBytes = 1;
+  uint32 maxObservationLengthBytes = 2;
+  uint32 maxReportLengthBytes = 3;
+  uint32 requestCountLimit = 4;
+  uint32 requestTotalBytesLimit = 5;
+  bool localRequest = 6;
```
Comment (bolekk) [inline go/ocr2/decryptionplugin/config/config_types.proto:18]: Maybe rename to "requireLocalRequestCheck" to make it clearer?

File: go/ocr2/decryptionplugin/decryption.go [hunk 7]
```diff
@@ -0,0 +1,375 @@
...
+			dp.logger.Error("DecryptionReporting Observation: cannot unmarshal and verify the ciphertext, the leader is faulty", commontypes.LogFields{
+				"error":        err,
+				"ciphertextID": request.CiphertextID,
+			})
+			return nil, fmt.Errorf("cannot unmarshal and verify the ciphertext: %w", err)
+		}
+		if dp.specificConfig.Config.LocalRequest {
+			queueCiphertextBytes, err := dp.decryptionQueue.GetCiphertext(request.CiphertextID)
+			if err != nil {
+				dp.logger.Error("DecryptionReporting Observation: cannot find ciphertext locally, skipping it", commontypes.LogFields{
```
[the remark this answers is not in the document]
Comment (bolekk) [reply go/ocr2/decryptionplugin/decryption.go:151]: I think regardless of the error, we have to skip the current ciphertext, so it doesn't really matter.

[the remark this answers is not in the document]
Comment (bolekk) [reply go/ocr2/decryptionplugin/decryption.go:151]: On a separate note: I'd lower log level to WARN or even INFO. I expect this to happen all the time (e.g. 1/4 nodes is slower and doesn't have the ciphertext ready locally). We don't want to flood logs with errors.

File: go/ocr2/decryptionplugin/decryption.go [hunk 8]
```diff
@@ -0,0 +1,375 @@
...
+				dp.logger.Error("DecryptionReporting Observation: local ciphertext does not match the query ciphertext, skipping it", commontypes.LogFields{
+					"ciphertextID": request.CiphertextID,
+				})
+				continue
+			}
+		}
+
+		decryptionShare, err := ciphertext.Decrypt(dp.privKeyShare)
+		if err != nil {
+			dp.logger.Error("DecryptionReporting Observation: cannot decrypt the ciphertext, the leader is faulty", commontypes.LogFields{
```
Comment (bolekk) [inline go/ocr2/decryptionplugin/decryption.go:166]: Let's remove "the leader is faulty" part. There could be other reasons for this failure.

File: go/ocr2/decryptionplugin/decryption.go [hunk 9]
```diff
@@ -0,0 +1,375 @@
...
+				continue
+			}
+
+			validDecryptionShares[ciphertextID] = append(validDecryptionShares[ciphertextID], validDecryptionShare)
+			if len(validDecryptionShares[ciphertextID]) >= fPlusOne {
+				dp.logger.Trace("DecryptionReporting Report: we have already f+1 valid decryption shares", commontypes.LogFields{
+					"ciphertextID": ciphertextID,
+					"observer":     ob.Observer,
+				})
+				break
```
Comment (bolekk) [inline go/ocr2/decryptionplugin/decryption.go:264]: Would it ever help to not break and collect more shares?

File: go/ocr2/decryptionplugin/queue.go [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+
+type DecryptionRequest struct {
+	ciphertextId CiphertextId
+	ciphertext   []byte
+}
+
+type DecryptionQueuingService interface {
+	GetRequests(requestCountLimit int, totalBytesLimit int) []DecryptionRequest
+	GetCiphertext(ciphertextId CiphertextId) ([]byte, error)
+	ReturnResult(ciphertextId CiphertextId, plaintext []byte)
```
Comment (bolekk) [inline go/ocr2/decryptionplugin/queue.go:13]: This was changed a couple times already but maybe "SetResult()" is a better name to follow the set/get pattern?

**Review by bolekk — APPROVED** [review]
This LGTM! Let's merge it, start integrating and iterate on it :)

File: go/ocr2/decryptionplugin/decryption.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (stchrysa) [reply go/ocr2/decryptionplugin/decryption.go:151]: @pszal good point.
@bolekk, @KuphJr should we have this method returning ([]byte, bool, error) and bool is whether the result was found or not? Wdyt?

[the remark this answers is not in the document]
Comment (pszal) [reply go/ocr2/decryptionplugin/decryption.go:151]: > @pszal good point. @bolekk, @KuphJr should we have this method returning ([]byte, bool, error) and bool is whether the result was found or not? Wdyt?

What about encoding this as `err`'s type? I.e., the method returns some type `ErrNotFound` if not found which is then checked with `if errors.Is(err, ErrNotFound)... `

File: go/ocr2/decryptionplugin/decryption.go (hunk 9 above)
Comment (stchrysa) [reply go/ocr2/decryptionplugin/decryption.go:264]: Only if we want to track which oracles sent valid shares.
But not for liveness, at this point we know that we have enough valid shares and that aggregation will succeed.

File: go/ocr2/decryptionplugin/queue.go (hunk 10 above)
Comment (stchrysa) [reply go/ocr2/decryptionplugin/queue.go:13]: SGTM, @KuphJr ?

File: go/ocr2/decryptionplugin/decryption.go (hunk 2 above)
Comment (stchrysa) [reply go/ocr2/decryptionplugin/decryption.go:272]: Yes, we have checked again [here](https://github.com/smartcontractkit/tdh2/blob/f0e8d0937917a332c1a3fb5c44798a6e76d24986/go/ocr2/decryptionplugin/decryption.go#LL238C1-L238C15).
It's a sanity check.

File: go/ocr2/decryptionplugin/decryption.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (bolekk) [reply go/ocr2/decryptionplugin/decryption.go]: I'd rather return a specific error, like pszal suggested. 
Still - is the behavior going to be any different? If we can't get the ciphertext from the local queue for whatever reason, we should skip processing.

[the remark this answers is not in the document]
Comment (pszal) [reply go/ocr2/decryptionplugin/decryption.go]: I think it makes sense to distinguish these cases for at least logging, e.g., not found -> INFO; faulty queue service -> ERROR. In the future, some retry mechanism could be added for the latter.

File: go/ocr2/decryptionplugin/queue.go (hunk 10 above)
Comment (KuphJr) [reply go/ocr2/decryptionplugin/queue.go:21]: Sure, I can make this change.

**Comment by bolekk** [comment]
Can we merge this PR, please? @pszal are you OK to approve or still expecting some changes? Have we agreed to add a special "ErrNotFound" error type?

**Review by pszal — APPROVED** [review]

**Comment by pszal** [comment]
> @pszal are you OK to approve or still expecting some changes?

I'm ok with merging and refining it later.

File: go/ocr2/decryptionplugin/decryption.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (stchrysa) [reply go/ocr2/decryptionplugin/decryption.go]: Done

**Review by kaleofduty — APPROVED** [review]
Let's merge for now and refine later.
Just to be clear: Consumers of this package should expect breaking changes to come out of further review/refinement.

File: .gitignore [hunk 11]
```diff
@@ -0,0 +1 @@
+go.work
```
Comment (kaleofduty) [inline .gitignore:1]: Don't we want this file in the repo?

**Review by KuphJr — APPROVED** [review]

File: .gitignore (hunk 11 above)
Comment (stchrysa) [reply .gitignore:1]: The tutorials suggest that each developer has their own file locally and that it's not added in the repository.

Comment (kaleofduty) [reply .gitignore:1]: I think we're in the "monorepo" case of https://github.com/golang/go/issues/53502 , i.e. we want to have one module for tdh2 crypto and another for the plugin, both in the same repo. So it seems that in our particular scenario we should check it in.

File: go/ocr2/decryptionplugin/types.proto [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+	repeated DecryptionShareWithID decryption_shares = 1;
+}
+
+message ProcessedDecryptionRequest {
+	bytes ciphertext_id = 1;
+	bytes plaintext = 2;
+}
+
+message Report {
+	repeated ProcessedDecryptionRequest processedDecryptedRequests = 1;
```
Comment (kaleofduty) [inline go/ocr2/decryptionplugin/types.proto:30]: does not follow style

**Pull request merged**